// @hubspot/api-client/lib/codegen/crm/companies/model/models.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./associatedId"));
__export(require("./batchInputSimplePublicObjectBatchInput"));
__export(require("./batchInputSimplePublicObjectId"));
__export(require("./batchInputSimplePublicObjectInput"));
__export(require("./batchReadInputSimplePublicObjectId"));
__export(require("./batchResponseSimplePublicObject"));
__export(require("./batchResponseSimplePublicObjectWithErrors"));
__export(require("./collectionResponseAssociatedId"));
__export(require("./collectionResponseSimplePublicObject"));
__export(require("./collectionResponseWithTotalSimplePublicObject"));
__export(require("./errorDetail"));
__export(require("./filter"));
__export(require("./filterGroup"));
__export(require("./modelError"));
__export(require("./nextPage"));
__export(require("./paging"));
__export(require("./publicObjectSearchRequest"));
__export(require("./simplePublicObject"));
__export(require("./simplePublicObjectBatchInput"));
__export(require("./simplePublicObjectId"));
__export(require("./simplePublicObjectInput"));
const associatedId_1 = require("./associatedId");
const batchInputSimplePublicObjectBatchInput_1 = require("./batchInputSimplePublicObjectBatchInput");
const batchInputSimplePublicObjectId_1 = require("./batchInputSimplePublicObjectId");
const batchInputSimplePublicObjectInput_1 = require("./batchInputSimplePublicObjectInput");
const batchReadInputSimplePublicObjectId_1 = require("./batchReadInputSimplePublicObjectId");
const batchResponseSimplePublicObject_1 = require("./batchResponseSimplePublicObject");
const batchResponseSimplePublicObjectWithErrors_1 = require("./batchResponseSimplePublicObjectWithErrors");
const collectionResponseAssociatedId_1 = require("./collectionResponseAssociatedId");
const collectionResponseSimplePublicObject_1 = require("./collectionResponseSimplePublicObject");
const collectionResponseWithTotalSimplePublicObject_1 = require("./collectionResponseWithTotalSimplePublicObject");
const errorDetail_1 = require("./errorDetail");
const filter_1 = require("./filter");
const filterGroup_1 = require("./filterGroup");
const modelError_1 = require("./modelError");
const nextPage_1 = require("./nextPage");
const paging_1 = require("./paging");
const publicObjectSearchRequest_1 = require("./publicObjectSearchRequest");
const simplePublicObject_1 = require("./simplePublicObject");
const simplePublicObjectBatchInput_1 = require("./simplePublicObjectBatchInput");
const simplePublicObjectId_1 = require("./simplePublicObjectId");
const simplePublicObjectInput_1 = require("./simplePublicObjectInput");
/* tslint:disable:no-unused-variable */
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
let enumsMap = {
    "BatchResponseSimplePublicObject.StatusEnum": batchResponseSimplePublicObject_1.BatchResponseSimplePublicObject.StatusEnum,
    "BatchResponseSimplePublicObjectWithErrors.StatusEnum": batchResponseSimplePublicObjectWithErrors_1.BatchResponseSimplePublicObjectWithErrors.StatusEnum,
    "Filter.OperatorEnum": filter_1.Filter.OperatorEnum,
};
let typeMap = {
    "AssociatedId": associatedId_1.AssociatedId,
    "BatchInputSimplePublicObjectBatchInput": batchInputSimplePublicObjectBatchInput_1.BatchInputSimplePublicObjectBatchInput,
    "BatchInputSimplePublicObjectId": batchInputSimplePublicObjectId_1.BatchInputSimplePublicObjectId,
    "BatchInputSimplePublicObjectInput": batchInputSimplePublicObjectInput_1.BatchInputSimplePublicObjectInput,
    "BatchReadInputSimplePublicObjectId": batchReadInputSimplePublicObjectId_1.BatchReadInputSimplePublicObjectId,
    "BatchResponseSimplePublicObject": batchResponseSimplePublicObject_1.BatchResponseSimplePublicObject,
    "BatchResponseSimplePublicObjectWithErrors": batchResponseSimplePublicObjectWithErrors_1.BatchResponseSimplePublicObjectWithErrors,
    "CollectionResponseAssociatedId": collectionResponseAssociatedId_1.CollectionResponseAssociatedId,
    "CollectionResponseSimplePublicObject": collectionResponseSimplePublicObject_1.CollectionResponseSimplePublicObject,
    "CollectionResponseWithTotalSimplePublicObject": collectionResponseWithTotalSimplePublicObject_1.CollectionResponseWithTotalSimplePublicObject,
    "ErrorDetail": errorDetail_1.ErrorDetail,
    "Filter": filter_1.Filter,
    "FilterGroup": filterGroup_1.FilterGroup,
    "ModelError": modelError_1.ModelError,
    "NextPage": nextPage_1.NextPage,
    "Paging": paging_1.Paging,
    "PublicObjectSearchRequest": publicObjectSearchRequest_1.PublicObjectSearchRequest,
    "SimplePublicObject": simplePublicObject_1.SimplePublicObject,
    "SimplePublicObjectBatchInput": simplePublicObjectBatchInput_1.SimplePublicObjectBatchInput,
    "SimplePublicObjectId": simplePublicObjectId_1.SimplePublicObjectId,
    "SimplePublicObjectInput": simplePublicObjectInput_1.SimplePublicObjectInput,
};
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap[expectedType]) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType; // w/e we don't know the type
            }
            // Check the discriminator
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType; // the type does not have a discriminator. use it.
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    if (typeMap[discriminatorType]) {
                        return discriminatorType; // use the type given in the discriminator
                    }
                    else {
                        return expectedType; // discriminator did not map to a type
                    }
                }
                else {
                    return expectedType; // discriminator was not present (or an empty string)
                }
            }
        }
    }
    static serialize(data, type) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) { // string.startsWith pre es6
            let subType = type.replace("Array<", ""); // Array<Type> => Type>
            subType = subType.substring(0, subType.length - 1); // Type> => Type
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.serialize(date, subType));
            }
            return transformedData;
        }
        else if (type === "Date") {
            return data.toISOString();
        }
        else {
            if (enumsMap[type]) {
                return data;
            }
            if (!typeMap[type]) { // in case we dont know the type
                return data;
            }
            // Get the actual type of this object
            type = this.findCorrectType(data, type);
            // get the map for the correct type.
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type);
            }
            return instance;
        }
    }
    static deserialize(data, type) {
        // polymorphism may change the actual type.
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) { // string.startsWith pre es6
            let subType = type.replace("Array<", ""); // Array<Type> => Type>
            subType = subType.substring(0, subType.length - 1); // Type> => Type
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.deserialize(date, subType));
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap[type]) { // is Enum
                return data;
            }
            if (!typeMap[type]) { // dont know the type
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.name] = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type);
            }
            return instance;
        }
    }
}
exports.ObjectSerializer = ObjectSerializer;
class HttpBasicAuth {
    constructor() {
        this.username = '';
        this.password = '';
    }
    applyToRequest(requestOptions) {
        requestOptions.auth = {
            username: this.username, password: this.password
        };
    }
}
exports.HttpBasicAuth = HttpBasicAuth;
class HttpBearerAuth {
    constructor() {
        this.accessToken = '';
    }
    applyToRequest(requestOptions) {
        if (requestOptions && requestOptions.headers) {
            const accessToken = typeof this.accessToken === 'function'
                ? this.accessToken()
                : this.accessToken;
            requestOptions.headers["Authorization"] = "Bearer " + accessToken;
        }
    }
}
exports.HttpBearerAuth = HttpBearerAuth;
class ApiKeyAuth {
    constructor(location, paramName) {
        this.location = location;
        this.paramName = paramName;
        this.apiKey = '';
    }
    applyToRequest(requestOptions) {
        if (this.location == "query") {
            requestOptions.qs[this.paramName] = this.apiKey;
        }
        else if (this.location == "header" && requestOptions && requestOptions.headers) {
            requestOptions.headers[this.paramName] = this.apiKey;
        }
        else if (this.location == 'cookie' && requestOptions && requestOptions.headers) {
            if (requestOptions.headers['Cookie']) {
                requestOptions.headers['Cookie'] += '; ' + this.paramName + '=' + encodeURIComponent(this.apiKey);
            }
            else {
                requestOptions.headers['Cookie'] = this.paramName + '=' + encodeURIComponent(this.apiKey);
            }
        }
    }
}
exports.ApiKeyAuth = ApiKeyAuth;
class OAuth {
    constructor() {
        this.accessToken = '';
    }
    applyToRequest(requestOptions) {
        if (requestOptions && requestOptions.headers) {
            requestOptions.headers["Authorization"] = "Bearer " + this.accessToken;
        }
    }
}
exports.OAuth = OAuth;
class VoidAuth {
    constructor() {
        this.username = '';
        this.password = '';
    }
    applyToRequest(_) {
        // Do nothing
    }
}
exports.VoidAuth = VoidAuth;
