// @hubspot/api-client/lib/codegen/crm/timelines/model/batchResponseTimelineEventResponse.js
"use strict";
/**
 * Timeline events
 * This feature allows an app to create and configure custom events that can show up in the timelines of certain CRM object like contacts, companies, or deals. You\'ll find multiple use cases for this API in the sections below.
 *
 * The version of the OpenAPI document: v3
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
* The state of the batch event request.
*/
class BatchResponseTimelineEventResponse {
    static getAttributeTypeMap() {
        return BatchResponseTimelineEventResponse.attributeTypeMap;
    }
}
exports.BatchResponseTimelineEventResponse = BatchResponseTimelineEventResponse;
BatchResponseTimelineEventResponse.discriminator = undefined;
BatchResponseTimelineEventResponse.attributeTypeMap = [
    {
        "name": "results",
        "baseName": "results",
        "type": "Array<TimelineEventResponse>"
    },
    {
        "name": "numErrors",
        "baseName": "numErrors",
        "type": "number"
    },
    {
        "name": "errors",
        "baseName": "errors",
        "type": "Array<Error>"
    },
    {
        "name": "status",
        "baseName": "status",
        "type": "BatchResponseTimelineEventResponse.StatusEnum"
    },
    {
        "name": "requestedAt",
        "baseName": "requestedAt",
        "type": "Date"
    },
    {
        "name": "startedAt",
        "baseName": "startedAt",
        "type": "Date"
    },
    {
        "name": "completedAt",
        "baseName": "completedAt",
        "type": "Date"
    }
];
(function (BatchResponseTimelineEventResponse) {
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["PENDING"] = 'PENDING'] = "PENDING";
        StatusEnum[StatusEnum["PROCESSING"] = 'PROCESSING'] = "PROCESSING";
        StatusEnum[StatusEnum["CANCELED"] = 'CANCELED'] = "CANCELED";
        StatusEnum[StatusEnum["COMPLETE"] = 'COMPLETE'] = "COMPLETE";
    })(StatusEnum = BatchResponseTimelineEventResponse.StatusEnum || (BatchResponseTimelineEventResponse.StatusEnum = {}));
})(BatchResponseTimelineEventResponse = exports.BatchResponseTimelineEventResponse || (exports.BatchResponseTimelineEventResponse = {}));
